### PR TITLE
New: no-useless-rename (fixes #6058)

### LINKED ISSUE #6058: Rule Proposal: no-useless-rename
From [disallowIdenticalDestructuringNames].

This rule will prevent to write useless renaming.
Original JSCS rule is checking only object patterns, but my proposal is checking object patterns, import declarations, and export declarations.

```js
{
    "no-useless-rename": ["error", {
        "ignoreDestructuring": false,
        "ignoreImport": false,
        "ignoreExport": false
    }]
}
```

**Valid:**

```js
let {foo: bar} = obj;
import * as foo from "foo";
import {foo as bar} from "foo";
export {foo as bar};
export {foo as bar} from "foo";
```

**Invalid:**

```js
let {foo: foo} = obj;
import {foo as foo} from "foo";
export {foo as foo};
export {foo as foo} from "foo";
```

[disallowIdenticalDestructuringNames]: http://jscs.info/rule/disallowIdenticalDestructuringNames

### PR DESCRIPTION
**Pull request by kaicataldo**

### CODE REVIEW

File: lib/rules/no-useless-rename.js [hunk 1]
```diff
@@ -0,0 +1,135 @@
...
+                i;
+
+            if (ignoreDestructuring) {
+                return;
+            }
+
+            properties = node.properties;
+
+            for (i = 0; i < properties.length; i++) {
+                if (properties[i].shorthand) {
```
Comment (mysticatea) [inline lib/rules/no-useless-rename.js]: I guess we should check `computed` or not.

File: lib/rules/no-useless-rename.js [hunk 2]
```diff
@@ -0,0 +1,135 @@
...
+                    return;
+                }
+
+                if (properties[i].key.name === properties[i].value.name) {
```
Comment (mysticatea) [inline lib/rules/no-useless-rename.js]: This `key` might be a `Literal`.

    let {"foo": foo} = {};

File: lib/rules/no-useless-rename.js [hunk 3]
```diff
@@ -0,0 +1,135 @@
...
+//------------------------------------------------------------------------------
+
+module.exports = {
+    meta: {
+        docs: {
+            description: "disallow renaming import, export, and destructured assignments to the same name",
+            category: "ECMAScript 6",
+            recommended: false
+        },
+        fixable: "code",
```
Comment (mysticatea) [inline lib/rules/no-useless-rename.js]: Yay! :sparkles:

File: docs/rules/no-useless-rename.md [hunk 4]
```diff
@@ -0,0 +1,116 @@
...
+```js
+/*eslint no-useless-rename: ["error", { ignoreDestructuring: true }]*/
+
+let { foo: foo } = bar;
+function foo({ bar: bar }) {}
+({ foo: foo }) => {}
+```
+
+## When Not To Use It
+
```
Comment (nzakas) [inline docs/rules/no-useless-rename.md]: Can you add a compatibility section that shows the JSCS equivalent? See http://eslint.org/docs/rules/no-octal#compatibility as an example.

File: lib/rules/no-useless-rename.js [hunk 5]
```diff
@@ -0,0 +1,142 @@
...
+                message: "{{type}} {{name}} unnecessarily renamed.",
+                data: {
+                    name: name,
+                    type: type
+                },
+                fix: function(fixer) {
+                    return fixer.replaceTextRange([
+                        initial.range[0],
+                        result.range[1]
+                    ], name);
```
Comment (kaicataldo) [inline lib/rules/no-useless-rename.js]: Replace the whole thing with just the name - this allows us to fix when keys are String Literals as well

File: lib/rules/no-useless-rename.js [hunk 6]
```diff
@@ -0,0 +1,142 @@
...
+            }
+
+            properties = node.properties;
+
+            for (i = 0; i < properties.length; i++) {
+                if (properties[i].shorthand) {
+                    return;
+                }
+
+                if (properties[i].computed) {
```
Comment (kaicataldo) [inline lib/rules/no-useless-rename.js]: Ignores computed properties like the original JSCS rule

File: lib/rules/no-useless-rename.js (hunk 2 above)
Comment (kaicataldo) [inline lib/rules/no-useless-rename.js]: Good call - updated!

**Comment by kaicataldo** [comment]
Updated :)

**Comment by mysticatea** [comment]
LGTM, Thank you!

**Comment by nzakas** [comment]
Nice work @kaicataldo!

**Pull request merged**